opsimulator/opsimulator.go

```diff
@@ -195,42 +196,126 @@ func (opSim *OpSimulator) handler(proxy *httputil.ReverseProxy, ctx context.Cont
 		r.Body = io.NopCloser(&buf)
 
 		// decode the fields we're interested in inspecting
-		msgs, err := readJsonMessages(body)
+		msgs, isBatchRequest, err := readJsonMessages(body)
 		if err != nil {
 			http.Error(w, fmt.Sprintf("Failed to parse JSON-RPC request: %s", err), http.StatusBadRequest)
 			return
 		}
 
-		for _, msg := range msgs {
+		rpcClient, err := rpc.Dial(opSim.l2Chain.Endpoint())
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Failed to connect to RPC server: %s", err), http.StatusInternalServerError)
+		}
+
+		batchRes := make([]rpcResponse, len(msgs))
+		for i, msg := range msgs {
 			if msg.Method == "eth_sendRawTransaction" {
 				var params []hexutil.Bytes
 				if err := json.Unmarshal(msg.Params, &params); err != nil {
-					http.Error(w, fmt.Sprintf("bad params sent to eth_sendRawTransaction: %s", err), http.StatusBadRequest)
-					return
+					opSim.log.Error(fmt.Sprintf("bad params sent to eth_sendRawTransaction: %s", err))
+					batchRes[i] = rpcResponse{
+						JSONRPC: "2.0",
+						Error: &jsonError{
+							Code:    -32000,
+							Message: err.Error(),
+						},
+						ID: msg.ID,
+					}
+					continue
 				}
 				if len(params) != 1 {
-					http.Error(w, "eth_sendRawTransaction request has invalid number of params", http.StatusBadRequest)
-					return
+					opSim.log.Error("eth_sendRawTransaction request has invalid number of params")
+					batchRes[i] = rpcResponse{
+						JSONRPC: "2.0",
+						Error: &jsonError{
+							Code:    -32000,
+							Message: "eth_sendRawTransaction request has invalid number of params",
+						},
+						ID: msg.ID,
+					}
+					continue
 				}
+
 				tx := new(types.Transaction)
 				if err := tx.UnmarshalBinary(params[0]); err != nil {
-					http.Error(w, fmt.Sprintf("failed to decode transaction data: %s", err), http.StatusBadRequest)
-					return
+					opSim.log.Error(fmt.Sprintf("failed to decode transaction data: %s", err))
+					batchRes[i] = rpcResponse{
+						JSONRPC: "2.0",
+						Error: &jsonError{
+							Code:    -32000,
+							Message: err.Error(),
+						},
+						ID: msg.ID,
+					}
+					continue
 				}
 
 				if err := opSim.checkInteropInvariants(ctx, tx); err != nil {
 					opSim.log.Error(fmt.Sprintf("interop invariants not met: %s", err))
-					// TODO (https://github.com/ethereum-optimism/supersim/issues/79) for batch requests write error to individual tx
-					http.Error(w, fmt.Sprintf("interop invariants not met: %s", err), http.StatusBadRequest)
-					return
+					batchRes[i] = rpcResponse{
+						JSONRPC: "2.0",
+						Error: &jsonError{
+							Code:    -32000,
+							Message: err.Error(),
+						},
+						ID: msg.ID,
+					}
+					continue
+				}
+			}
+
+			batchRes[i], err = forwardRPCRequest(ctx, rpcClient, msg)
+			if err != nil {
+				opSim.log.Error(fmt.Sprintf("RPC returned an error: %s", err))
+				batchRes[i] = rpcResponse{
+					JSONRPC: "2.0",
+					Error: &jsonError{
+						Code:    -32000,
+						Message: err.Error(),
```

Comment:
might be DRY to make a helper function for instantiating the error object

Reply:
hmm, I dont think we get much out of creating a helper for that